TPP.Persistence.MongoDB/Repos/BadgeMarketRepo.cs

```diff
@@ -0,0 +1,249 @@
+                cm.MapProperty(o => o.Source).SetElementName("source");
+                cm.MapProperty(o => o.CreatedAt).SetElementName("created_at");
+                cm.MapProperty(o => o.Price).SetElementName("price");
+                cm.MapProperty(o => o.Amount).SetElementName("amount");
+                cm.MapProperty(o => o.Form).SetElementName("form")
+                   .SetIgnoreIfNull(true);
+                cm.MapProperty(o => o.Shiny).SetElementName("shiny")
+                    .SetDefaultValue(false)
+                    .SetIgnoreIfDefault(true);
+            });
+        }
+
+        private void InitIndexes()
+        {
+            BuyOfferCollection.Indexes.CreateMany(new[]
+            {
+                new CreateIndexModel<BadgeBuyOffer>(Builders<BadgeBuyOffer>.IndexKeys.Ascending(u => u.UserId)),
+                new CreateIndexModel<BadgeBuyOffer>(Builders<BadgeBuyOffer>.IndexKeys.Ascending(u => u.Species)),
+                new CreateIndexModel<BadgeBuyOffer>(Builders<BadgeBuyOffer>.IndexKeys.Ascending(u => u.Source)),
+                new CreateIndexModel<BadgeBuyOffer>(Builders<BadgeBuyOffer>.IndexKeys.Ascending(u => u.CreatedAt)),
+                new CreateIndexModel<BadgeBuyOffer>(Builders<BadgeBuyOffer>.IndexKeys.Ascending(u => u.Form)),
+                new CreateIndexModel<BadgeBuyOffer>(Builders<BadgeBuyOffer>.IndexKeys.Ascending(u => u.Shiny)),
+            });
+        }
+
+        public BadgeMarketRepo(IMongoDatabase database, IBadgeRepo badgeRepo, IUserRepo userRepo, IBank<User> bank, IClock clock)
+        {
+            _badgeRepo = badgeRepo;
+            _userRepo = userRepo;
+            _tokenBank = bank;
+            database.CreateCollectionIfNotExists(BuyOfferCollectionName).Wait();
+            BuyOfferCollection = database.GetCollection<BadgeBuyOffer>(BuyOfferCollectionName);
+            _clock = clock;
+            InitIndexes();
+        }
+
+        public async Task<List<BadgeBuyOffer>> FindAllBuyOffers(string? userId, PkmnSpecies species, string? form, Badge.BadgeSource? source, bool? shiny)
+        {
+            FilterDefinition<BadgeBuyOffer> filter = Builders<BadgeBuyOffer>.Filter.Empty;
+            if (userId != null)
+                filter &= Builders<BadgeBuyOffer>.Filter.Eq(b => b.UserId, userId);
+            if (species != null)
+                filter &= Builders<BadgeBuyOffer>.Filter.Eq(b => b.Species, species);
+            if (form != null)
+                filter &= Builders<BadgeBuyOffer>.Filter.Eq(b => b.Form, form);
+            if (source != null)
+                filter &= Builders<BadgeBuyOffer>.Filter.Eq(b => b.Source, source);
+            if (shiny == true)
+                filter &= Builders<BadgeBuyOffer>.Filter.Eq(b => b.Shiny, true);
+            else
+                filter &= Builders<BadgeBuyOffer>.Filter.Ne(b => b.Shiny, true);
+
+            return await BuyOfferCollection.Find(filter).ToListAsync();
+        }
```

Comment:
hmm not sure how I feel about this duplicate code. Especially since it is a bit finnicky in regards to the shiny filter requiring to filter for "!= true". If this is equal to the variant in BadgeRepo, can it be extracted as a utility method somewhere that builds a `FilterDefinition<T>`?